Reduce allocations in System.Threading.Channels for netstandard2.* and net461

KorneiDontsov opened the pull request:
`System.Threading.Channels.AsyncOperation<TResult>` is well optimized for .NET Core, but there are no optimizations for .NET Standard 2.0 and .NET Framework. They use the same code as .NET Standard 1.3.
I used ThreadPool and ExecutionContext API, that are available in .NET Standard 2.* and .NET Framework, to reduce the amount of allocations.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=46458) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=46458) before we can accept your contribution.<br/><br/>:x: KorneiDontsov [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=46458)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=46458) it.</sub>

KorneiDontsov committed:
Reduce allocations in System.Threading.Channels for netstandard2.* and net461

System.Threading.Channels.AsyncOperation<TResult> is well optimized for .NET Core, but there are no optimizations for .NET Standard 2.0 and .NET Framework. They use the same code as .NET Standard 1.3.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.netstandard2.cs:
```diff
@@ -0,0 +1,27 @@
+            using (ExecutionContext.SuppressFlow())
+            {
+                return cancellationToken.Register(action, state);
+            }
```
Can you share your perf test / results that show this ExecutionContext-related code actually improves performance?

stephentoub replied:
@KorneiDontsov, are you still working on this?  Thanks.

KorneiDontsov replied:
@stephentoub, sorry that I didn't answer before - I had a weekend.

I'll look at the benchmarks on dotnet/performance in a few days. If I'll find that some issues are not tested, i'll add the benchmarks, required to see the difference in allocations.

I used netstandard2.0 build of System.Threading.Channels in Unity Engine project at work and tested it in Unity Profiler. I found that I get ExecutionContext.Capture() invocation with 40B allocation every time I write item to bounded channel. I'm pretty sure I can reproduce the behaviour on .NET Framework and share you the results.

KorneiDontsov replied:
@stephentoub, I found out that it's not that easy to benchmark over `net461`, so I just put `netstandard2.0` variation of System.Threading.Channels into corerun and test it over `netcoreapp5.0`. The commands I used:

```
dotnet-runtime> .\build.cmd clr+libs -rc Release -c Release
dotnet-runtime> .\build.cmd -projects .\src\libraries\System.Threading.Channels\System.Threading.Channels.sln -c Release -f netstandard2.0
dotnet-performance> py .\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter System.Threading.Channels* --corerun "..\dotnet-runtime\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.0\CoreRun.exe" --bdn-arguments="--artifacts "C:\dotnet-artifacts\benchmarks\channels-origin""
```

I added one benchmark of ping-pong with non-default execution context to find any differences. Take a look at https://github.com/dotnet/performance/pull/1634.

I benchmarked 3 times:
- `channels-origin` -- without any changes
- `channels-full-optimized` --with all changes
- `channels-optimized-but-safe-register` -- with all changes except suppression of execution flow in `AsyncOperation<T>.UnsafeRegister`.

The results are pretty same for all channels so I show the results for unbounded channel. Also I attached all results: [benchmarks.zip](https://github.com/dotnet/runtime/files/5833071/benchmarks.zip).

### origin

|                        Method |             Mean |          Error |         StdDev |           Median |              Min |              Max |    Gen 0 | Gen 1 | Gen 2 |   Allocated |
|------------------------------ |-----------------:|---------------:|---------------:|-----------------:|-----------------:|-----------------:|---------:|------:|------:|------------:|
|           TryWriteThenTryRead |         70.94 ns |       0.160 ns |       0.149 ns |         70.90 ns |         70.76 ns |         71.23 ns |        - |     - |     - |           - |
|       WriteAsyncThenReadAsync |        121.77 ns |       0.529 ns |       0.469 ns |        121.67 ns |        121.11 ns |        122.83 ns |        - |     - |     - |           - |
|       ReadAsyncThenWriteAsync |        183.28 ns |       0.419 ns |       0.371 ns |        183.15 ns |        182.85 ns |        184.08 ns |        - |     - |     - |           - |
|                      PingPong | 39,969,046.43 ns | 242,068.034 ns | 214,587.007 ns | 39,949,500.00 ns | 39,674,800.00 ns | 40,303,300.00 ns | 250.0000 |     - |     - | 1,267,408 B |
| PingPongWithExecutionContexts | 41,516,956.67 ns | 369,187.887 ns | 345,338.578 ns | 41,549,450.00 ns | 40,764,500.00 ns | 42,002,875.00 ns | 500.0000 |     - |     - | 2,861,484 B |

### optimized-but-safe-register

|                        Method |             Mean |          Error |         StdDev |           Median |              Min |              Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |-----------------:|---------------:|---------------:|-----------------:|-----------------:|-----------------:|------:|------:|------:|----------:|
|           TryWriteThenTryRead |         71.07 ns |       0.150 ns |       0.140 ns |         71.03 ns |         70.78 ns |         71.31 ns |     - |     - |     - |         - |
|       WriteAsyncThenReadAsync |        122.51 ns |       0.311 ns |       0.260 ns |        122.47 ns |        122.03 ns |        123.14 ns |     - |     - |     - |         - |
|       ReadAsyncThenWriteAsync |        184.76 ns |       0.357 ns |       0.334 ns |        184.68 ns |        184.32 ns |        185.31 ns |     - |     - |     - |         - |
|                      PingPong | 33,965,764.44 ns | 175,180.917 ns | 163,864.338 ns | 33,938,983.33 ns | 33,665,000.00 ns | 34,249,166.67 ns |     - |     - |     - | 649,560 B |
| PingPongWithExecutionContexts | 34,050,685.56 ns | 233,693.254 ns | 218,596.815 ns | 34,143,083.33 ns | 33,669,666.67 ns | 34,316,816.67 ns |     - |     - |     - | 656,399 B |

### full-optimized

|                        Method |             Mean |          Error |         StdDev |           Median |              Min |              Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |-----------------:|---------------:|---------------:|-----------------:|-----------------:|-----------------:|------:|------:|------:|----------:|
|           TryWriteThenTryRead |         72.47 ns |       0.142 ns |       0.126 ns |         72.43 ns |         72.32 ns |         72.71 ns |     - |     - |     - |         - |
|       WriteAsyncThenReadAsync |        121.94 ns |       0.247 ns |       0.219 ns |        121.98 ns |        121.44 ns |        122.19 ns |     - |     - |     - |         - |
|       ReadAsyncThenWriteAsync |        187.86 ns |       0.209 ns |       0.185 ns |        187.79 ns |        187.63 ns |        188.16 ns |     - |     - |     - |         - |
|                      PingPong | 34,224,820.24 ns | 166,505.201 ns | 147,602.524 ns | 34,245,800.00 ns | 33,914,600.00 ns | 34,515,483.33 ns |     - |     - |     - | 652,792 B |
| PingPongWithExecutionContexts | 34,290,106.67 ns | 345,391.007 ns | 323,078.962 ns | 34,331,166.67 ns | 33,624,750.00 ns | 34,701,550.00 ns |     - |     - |     - | 652,981 B |

## Conclution

1. `full-optimized` has no advantages over `optimized-but-safe-register`. So you're right: execution flow suppression is redundant. I removed it in the next commit.
2.  Benchmark `PingPongWithExecutionContexts` in `origin` has 2,3 times more allocations then `PingPong`. However, this benchmark has no extra allocations in `optimized-but-safe-register` and `full-optimized`. So, I highly recommend to commit this benchmark to [dotnet/performance](https://github.com/dotnet/performance).

## ResultsComparer

Also take a look at the execution time differences.

```
dotnet-performance\src\tools\ResultsComparer> dotnet run --base "C:\dotnet-artifacts\benchmarks\channels-origin\" --diff "C:\dotnet-artifacts\benchmarks\channels-optimized-but-safe-register\" --threshold 1%
```

summary:
better: 8, geomean: 1.138
worse: 1, geomean: 1.018
total diff: 9

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.WriteAsyncThenRead |      1.02 |           106.18 |           108.06 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Threading.Channels.Tests.UnboundedChannelPerfTests.PingPongWithExecutionC |      1.22 |      41549450.00 |      34143083.33 |         |
| System.Threading.Channels.Tests.BoundedChannelPerfTests.PingPongWithExecutionCon |      1.20 |      41163287.50 |      34389225.00 |         |
| System.Threading.Channels.Tests.BoundedChannelPerfTests.PingPong                 |      1.18 |      40097075.00 |      33905183.33 |         |
| System.Threading.Channels.Tests.UnboundedChannelPerfTests.PingPong               |      1.18 |      39949500.00 |      33938983.33 |         |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.PingPongWithExecut |      1.12 |      40383237.50 |      35953308.33 | bimodal |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.PingPong           |      1.10 |      39509325.00 |      35921891.67 | several?|
| System.Threading.Channels.Tests.BoundedChannelPerfTests.WriteAsyncThenReadAsync  |      1.09 |           147.23 |           135.16 |         |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.ReadAsyncThenWrite |      1.03 |           192.11 |           185.73 |         |

KorneiDontsov committed:
Remove execution flow suppression as redundant

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.netstandard2.cs:
```diff
@@ -0,0 +1,17 @@
+            ThreadPool.UnsafeQueueUserWorkItem(static s => ((AsyncOperation<TResult>)s).SetCompletionAndInvokeContinuation(), this);
+
+        private static void QueueUserWorkItem(Action<object?> action, object? state) =>
+            ThreadPool.QueueUserWorkItem(new WaitCallback(action), state);
```
Have you measured reverting this QueueUserWorkItem back to using Task.Factory.StartNew?  I suspect the previous code for this is a tad better.

KorneiDontsov replied:
[QueueUserWorkItemOptimizationBenchmarks.zip](https://github.com/dotnet/runtime/files/5835206/QueueUserWorkItemOptimizationBenchmarks.zip)

The performance with and without change is quite the same with threshold 4%, but I found very small differences in allocations.

| Method | with | without |
| --------- |-----:|---------:|
| BoundedChannelPerfTests.PingPong | 645,197 B | 635,464 B |
| BoundedChannelPerfTests.PingPongWithExecutionContexts | 652,358 B | 667,669 B |
| SpscUnboundedChannelPerfTests.PingPong | 641,473 B | 633,981 B |
| SpscUnboundedChannelPerfTests.PingPongWithExecutionContexts | 645,224 B | 666,235 B |
| UnboundedChannelPerfTests.PingPong | 647,304 B | 637,213 B |
| UnboundedChannelPerfTests.PingPongWithExecutionContexts | 651,749 B | 674,043 B |

Task.Factory.StartNew is better with default ExecutionContext and worse with custom. PingPong has 10 000 iterations, so we have a difference of 1-3 bytes per iteration. It's hard for me to say which implementation is better. @stephentoub, what do you think?

stephentoub replied:
> @stephentoub, what do you think?

Let's stick with Task.Factory.StartNew.  On .NET Framework, modifying ExecutionContext means that every subsequent flowing incurs allocation, so most components then optimize for the non-modified case, including Task/StartNew.  This can do the same.

KorneiDontsov replied:
Done.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.netstandard2.cs:
```diff
@@ -0,0 +1,17 @@
+    internal partial class AsyncOperation<TResult>
+    {
+        private void UnsafeQueueSetCompletionAndInvokeContinuation() =>
+            ThreadPool.UnsafeQueueUserWorkItem(static s => ((AsyncOperation<TResult>)s).SetCompletionAndInvokeContinuation(), this);
```
This one makes sense.

KorneiDontsov replied:
Yes. After all measurements, it is clear that this change is the only one that really optimizes memory and performance.

KorneiDontsov committed:
Rollback QueueUserWorkItem

stephentoub reviewed (approved):
Thanks.

KorneiDontsov commented:
@stephentoub, you're welcome!

KorneiDontsov commented:
Some checks were not successful. Why? Can I help with this?

danmosemsft commented:
I hit retry on those jobs.

danmosemsft commented:
Thanks for the contribution @KorneiDontsov . We welcome contributions, perhaps there is some other issue or contribution you're interested in in.
